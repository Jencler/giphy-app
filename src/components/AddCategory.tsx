import { useState, type ChangeEvent, type FormEvent } from "react";

interface Props {
  onAddCategory: (categoryName: string) => void;
}

export const AddCategory = ({ onAddCategory }: Props) => {
  const [categoryName, setCategoryName] = useState("");

  const handleChange = ({ target }: ChangeEvent<HTMLInputElement>) => {
    setCategoryName(target.value);
  };

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const trimmed = categoryName.trim();
    if (trimmed === "" || trimmed.length <= 1) return;
    if (!/^[a-zA-Z\s]+$/.test(trimmed)) return;
    onAddCategory(trimmed);
    setCategoryName("");
  };

  return (
    <section className="main_middle-section">
      <form role="form" aria-label="form" onSubmit={handleSubmit}>
        <label className="label">
          <span>Agrega una categor√≠a</span>
          <input
            type="text"
            placeholder="Happy, smile, love, ..."
            value={categoryName}
            onChange={handleChange}
            autoComplete="off"
            autoFocus
          />
        </label>
      </form>
    </section>
  );
};
