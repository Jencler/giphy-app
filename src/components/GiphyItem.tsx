import type { Gifs } from "../helpers/getGifs";

export const GiphyItem = ({ title, image }: Gifs) => {
  return (
    <article className="card-giphy">
      <h3>{title}</h3>
      <img src={image} alt={title} />
    </article>
  );
};
