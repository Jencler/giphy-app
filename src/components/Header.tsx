import { Logo } from "./Logo";

export const Header = () => {
  return (
    <header className="header">
      <Logo />
    </header>
  );
};
