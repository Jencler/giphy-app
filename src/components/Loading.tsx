export const Loading = () => {
  return (
    <div data-testid="loader" className="spinner">
      <div className="spinnerin"></div>
    </div>
  );
};
