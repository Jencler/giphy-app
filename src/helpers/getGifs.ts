import type { Datum } from "../interfaces/giphys.interfaces";

export interface Gifs {
  id: string;
  title: string;
  image: string;
}
const API_KEY = "WXM03H9JcLJFMDG76xZeapiklSjBzmDG";
const LIMIT = 10;

export const getGiphy = async (category: string): Promise<Gifs[]> => {
  try {
    const response = await fetch(
      `https://api.giphy.com/v1/gifs/search?q=${category}&api_key=${API_KEY}&limit=${LIMIT}`
    );
    const { data } = await response.json();

    const gifs: Gifs[] = data.map((gif: Datum) => ({
      id: gif.id,
      title: gif.title,
      image: gif.images.downsized_medium.url,
    }));

    return gifs;
  } catch (error) {
    console.error("getGiphy error", error);
    return [];
  }
};
