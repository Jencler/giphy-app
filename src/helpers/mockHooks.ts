export function mockHook<T extends (...args: any[]) => any>(
  hook: T
): jest.Mock {
  return hook as unknown as jest.Mock;
}
