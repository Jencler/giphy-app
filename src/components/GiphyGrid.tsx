import { GiphyItem } from "./GiphyItem";
import { Loading } from "./Loading";
import { useFetchingGif } from "../hooks/useFetchingGif";

interface Props {
  category: string;
}

export const GiphyGrid = ({ category }: Props) => {
  const { gifs, isLoading } = useFetchingGif({ category });

  return (
    <>
      {isLoading ? (
        <Loading />
      ) : (
        <div className="cards-containers">
          <h2 className="subtitle">{category}</h2>
          <section className="grid-giphys">
            {gifs.map((giphy) => (
              <GiphyItem key={giphy.id} {...giphy} />
            ))}
          </section>
        </div>
      )}
    </>
  );
};
