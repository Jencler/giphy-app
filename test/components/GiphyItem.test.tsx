import { render, screen } from "@testing-library/react";
import { GiphyItem } from "../../src/components/GiphyItem";
import { Gifs } from "../../src/helpers/getGifs";

describe("Componente GiphyItem", () => {
  const gif: Gifs = {
    id: "cjudF7TsDnKsqTVxAL",
    title: "Disney Pixar GIF by Disney+",
    image:
      "https://media3.giphy.com/media/cjudF7TsDnKsqTVxAL/giphy.gif?cid=58ff979es5to6cuh2fwn99qxgl2wib8n2vmaevaye42ne3hw&ep=v1_gifs_search&rid=giphy.gif&ct=g",
  };

  test("Debe de hacer match con el snapshot", () => {
    const { container } = render(<GiphyItem {...gif} />);
    expect(container).toMatchSnapshot();
  });

  test("Debe de mostrar la imagen con el URL y el ALT indicado", () => {
    render(<GiphyItem {...gif} />);
    const { src, alt } = screen.getByRole("img") as HTMLImageElement;
    expect(screen.getByAltText(gif.title)).toBeInTheDocument();
    expect(src).toBe(gif.image);
    expect(alt).toBe(gif.title);
  });

  test("Debe de mostrar el titulo en el componente", () => {
    render(<GiphyItem {...gif} />);
    expect(screen.getByText(gif.title)).toBeTruthy();
  });
});
