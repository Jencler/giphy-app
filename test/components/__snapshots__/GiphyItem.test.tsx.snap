// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente GiphyItem Debe de hacer match con el snapshot 1`] = `
<div>
  <article
    class="card-giphy"
  >
    <h3>
      Disney Pixar GIF by Disney+
    </h3>
    <img
      alt="Disney Pixar GIF by Disney+"
      src="https://media3.giphy.com/media/cjudF7TsDnKsqTVxAL/giphy.gif?cid=58ff979es5to6cuh2fwn99qxgl2wib8n2vmaevaye42ne3hw&ep=v1_gifs_search&rid=giphy.gif&ct=g"
    />
  </article>
</div>
`;
